lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -381,19 +490,190 @@ instance ToText FollowLog where
+
+--
+-- Log aggregation
+--
+
+-- | Statistics of interest from the follow-function.
+--
+-- The @f@ allows us to use @LogState@ to keep track of both current and
+-- previously logged stats, and perform operations over it in a nice way.
+data FollowStats f = FollowStats
+    { blocksApplied :: f Int
+    , rollbacks :: f Int
+    , time :: f UTCTime
+      -- ^ NOTE: Current time is not updated until @flush@ is called.
+    , prog :: f SyncProgress
+    } deriving (Generic)
+
+deriving instance Show (FollowStats LogState)
+deriving instance Eq (FollowStats LogState)
+deriving instance NFData (FollowStats LogState)
+
+-- | Change the @f@ wrapping each record field.
+hoistStats
+    :: (forall a. f a -> g a)
+    -> FollowStats f
+    -> FollowStats g
+hoistStats f FollowStats{blocksApplied,rollbacks,time,prog} = FollowStats
+    { blocksApplied = f blocksApplied
+    , rollbacks = f rollbacks
+    , time = f time
+    , prog = f prog
+    }
+
+-- | For keeping track of what we have logged and what we have not.
+--
+-- The idea is to
+-- 1. Reconstruct a model of the @current@ @state@ using a @Trace@
+-- 2. Sometimes log the difference between the @current@ state and the most
+-- recently logged one.
+data LogState a = LogState
+    { prev :: a -- ^ Most previously logged state
+    , current :: a -- ^ Not-yet logged state
+    } deriving (Eq, Show, Functor, Generic)
+
+deriving instance NFData a => NFData (LogState a)
+
+initLogState :: a -> LogState a
+initLogState a = LogState a a
+
+-- | Modify the current state of a @LogState state@
+overCurrent :: (a -> a) -> LogState a -> LogState a
+overCurrent f (LogState prev cur) = LogState prev (f cur)
+
+-- | /The way/ to log the current stats.
+--
+-- It /traces/ a presentable @FollowStats LogState@ before flushing and
+-- /returns/ a new flushed @FollowStats@ where @prev == current@.
+flush
+    :: Monad m
+    => Tracer m (FollowStats LogState)
+    -> UTCTime
+    -> FollowStats LogState
+    -> m (FollowStats LogState)
+flush tr t s = do
+    -- This is where we need to update the time
+    let s' = s { time = overCurrent (const t) (time s) }
+    traceWith tr s'
+    return $ hoistStats forgetPrev s'
+  where
+    forgetPrev (LogState _prev cur) = LogState cur cur
+
+emptyStats :: UTCTime -> SyncProgress -> FollowStats LogState
+emptyStats t p = FollowStats (f 0) (f 0) (f t) (f p)
+  where
+    f = initLogState
+
+-- | Update the stats based on a new log message.
+updateStats :: FollowLog tr -> FollowStats LogState -> FollowStats LogState
+updateStats msg s = force $ case msg of
+    MsgApplyBlocks _tip blocks ->
+        s { blocksApplied = overCurrent (+ NE.length blocks) (blocksApplied s) }
+    MsgDidRollback _ _ ->
+        s { rollbacks = overCurrent (1 +) (rollbacks s) }
+    MsgSyncProgress p ->
+        s { prog = overCurrent (const p) (prog s) }
+    _ -> s
+
+instance ToText (FollowStats LogState) where
+    toText (FollowStats b r t sp) = syncStatus <> " " <> stats
```

Comment:
This time-based summary is good for the stake pools thread, but for wallets I think it should be logging when there are incoming transactions, or when restoration is in progress, and complete silence otherwise.

I'll move this statistics code into the Cardano.Wallet.Logging module too.

Reply:
> but for wallets I think it should be logging when there are incoming transactions

Just to be clear: the logs collected for aggregation are still shown or not depending on severity. It is still logging discoveries.

> or when restoration is in progress

Yes, ideally the `SyncProgress` `LogState` transitions should probably be logged immediately when they occur, not only from a periodic monitor-check. My hunch is still that regular status updates wallet workers are worth it too, though — a summary every 5 minutes is likely small compared to other Info logs. For instance, Daedalus seems to be polling a lot, appearing as request logs.

Reply:
Or I guess we could  _monitor_ / poll the progress regularly without _logging_ it (+ stats), if we are still in-sync.

If the chain-following were to halt, we'd then notice, without being _required_ to log every-five-minutes status updates.